test/resource/attributes_test.exs

```diff
@@ -91,5 +91,19 @@ defmodule Ash.Test.Resource.AttributesTest do
         end
       )
     end
+
+    test "raises if you pass a reserved name for `name`" do
```

Comment:
This should be considered unsatisfactory testing (we of course need to test for all of the reserved names). 

I wanted to iterate over all reserved names in a for-comprehension, but I had trouble getting my tests to work as I wanted to.

It seems like I need to create a new module for each iteration, as otherwise I'll get
```elixir
     test/resource/attributes_test.exs:23
     ** (ArgumentError) field/association :__metadata__ is already set on schema
     stacktrace:
       (ecto 3.7.2) lib/ecto/schema.ex:2146: Ecto.Schema.put_struct_field/3
       (ecto 3.7.2) lib/ecto/schema.ex:1899: Ecto.Schema.define_field/4
       (stdlib 3.17.1) erl_eval.erl:685: :erl_eval.do_apply/6
       (stdlib 3.17.1) erl_eval.erl:123: :erl_eval.exprs/5
       (stdlib 3.17.1) erl_eval.erl:919: :erl_eval.try_clauses/8
       (stdlib 3.17.1) erl_eval.erl:123: :erl_eval.exprs/5
       (elixir 1.13.0) src/elixir.erl:289: :elixir.recur_eval/3
       (spark 0.1.25) /home/juha/projects/elixir/ash/test/resource/attributes_test.exs:24: Spark.Dsl.__before_compile__/1
```

However, despite my attempts, I couldn't get this to work right (yet).

Reply:
Yeah, I for loop would be ideal, but its not too big of a deal

Reply:
I'd also prefer us to have a neat loop, so I dug a bit more!

I pushed a new commit to demonstrate the problem I'm having, and I figured this might be the best avenue to ask if you have other ideas. 

As you can see in the latest commit, I identified my "problem" to be the return value of the `after_compile?` function. By setting it to `false`, I was able to sidestep my issue where otherwise I'd get the "is already set on schema" ArgumentError. 

As it seems to me that we want the return value to remain `true`, I want to ask: can you come up with a workaround for this case, or explain why specifically I might be running into this?

Reply:
Yeah, okay. So the issue is that `after_compile?` transformers happen after we build the ecto schema, which is a pretty interesting problem. I think we want to retain the nice error messages, so what we ought to do is just ignore any overlaps in the schema. What we can do to make that happen is probably here:

https://github.com/ash-project/ash/blob/main/lib/ash/resource/schema.ex#L92

And just say `struct_fields = Module.get_attribute(__MODULE__, :ash_struct_fields) |> Enum.reject(<is_reserved>)`

Reply:
That same code exists in the other branch of that `if` so it has to be done there as well.

Reply:
Thanks Zach! I think I figured this one out, and it doesn't seem to have been about that particular line after all (unless I'm mistaken, of course 😄).

If you have a look at my latest commit, you'll see that I now filter the attributes out at the very beginning, before we call to `Ecto.Schema.field/3`. I realized that we were doing that against each user-provided attribute, and then attempting to call `Ecto.Schema.field/3` for some of the reserved field names right below. This obviously didn't work, and makes perfect sense in the light of the error I was seeing.

What I do still want to fix about this PR is the scattering of knowledge: now I'm just hard-coding the list of reserved names in `lib/ash/resource/schema.ex`, but ideally we'd have a single source to reference. However, I don't have quite the mental model of Ash to best judge where to place such information so that both `lib/ash/resource/schema.ex` and `lib/ash/resource/transformers/no_reserved_field_names.ex` could reference the same source, as well as the test at `test/resource/attributes_test.exs`. Any suggestions on this?

Reply:
That makes sense, but you will need to do the same for aggregates, relationships, and calculations in both branches as well :)

In terms of where the list lives, I think its fine to make it public and putting it in `Ash.Resource` sounds like a good idea :)

Reply:
Thanks! 

I pushed an update to relocate the list of reserved names, but I wasn't quite sure about your suggestion on "do the same for aggregates, relationships, and calculations in both branches as well". The PR currently does this for both branches (for both `schema` and `embedded_schema`), and the fields for aggregates, relationships, and calculations seem to be within these two schemas. Are you referring to something else that I'm not getting here?

I also saw that Dialyzer is failing, but that seems to be true for the master branch as well.

Reply:
https://github.com/ash-project/ash/pull/388/files#diff-0444918c54e2c63a6c07cd12fcd2da2e8e33119c56afccbf57d5a2ce3131d66cL156

When building the schema, we add virtual fields for aggregates and calculations, so those would need to be updated as well not to use reserved names

Reply:
Sorry, here is a better link: https://github.com/ash-project/ash/blob/main/lib/ash/resource/schema.ex#L68